pontoon/base/models.py

```diff
@@ -1721,179 +1721,256 @@ def __str__(self):
-                            translation__fuzzy=True,
-                            translation__approved=False,
-                            translation__locale=locale
-                        ), then=1
-                    ), output_field=models.IntegerField(), default=0
-                )
-            ),
-            suggested_count=Sum(
-                Case(
-                    When(
-                        Q(
-                            translation__locale=locale,
-                            translation__approved=False,
-                            translation__fuzzy=False,
-                        ),
-                        then=1
-                    ),
-                    output_field=models.IntegerField(),
-                    default=0
-                )
-            ),
-            unreviewed_count=Sum(
-                Case(
-                    When(
-                        Q(
-                            translation__locale=locale,
-                            translation__approved=False,
-                            translation__fuzzy=False,
-                            translation__rejected=False,
-                        ),
-                        then=1
+    def get_filtered_entities(self, locale, query, rule, match_all=True):
+        """Return a QuerySet of values of entity PKs matching the locale, query and rule.
+
+        Filter entities that match the given filter provided by the `locale` and `query`
+        parameteres. For performance reasons the `rule` parameter is also provided to filter
+        entities in python instead of the DB.
+
+        :arg Locale locale: a Locale object to get translations for
+        :arg Q query: a django ORM Q() object describing translations to filter
+        :arg function rule: a lambda function implementing the `query` logic
+        :arg boolean match_all: if true, all plural forms must match the rule.
+            Otherwise, only one matching is enough
+
+        :returns: a QuerySet of values of entity PKs
+
+        """
+        # First of find all translations that match the criteria.
+        translations = Translation.objects.filter(locale=locale).filter(query)
+
+        plural_pks = []
+
+        if locale.nplurals:
+            # Then we want to find the active translation of each plural form of each
+            # entity that has plurals.
+            # So we query all those entities, with for each a list of translations ordered
+            # so that the first one for each plural form will be the active one.
+            plural_candidates = (
+                self
+                .exclude(string_plural='')
+                .filter(pk__in=translations.values('entity'))
+                .prefetch_related(Prefetch(
+                    'translation_set',
+                    queryset=(
+                        Translation.objects
+                        .filter(locale=locale)
+                        .order_by('approved', 'fuzzy', '-date')
                     ),
```

Comment:
I think `I think `queryset=translations.order_by('approved', 'fuzzy', '-date')` would work here. And then you don't need `.filter(pk__in=translations.values('entity'))` above.